stat_plan9.go

```diff
@@ -42,63 +44,53 @@ func translateSyscallError(err error) (uint32, bool) {
 }
 
 // toFileMode converts sftp filemode bits to the os.FileMode specification
-func toFileMode(mode uint32) os.FileMode {
-	var fm = os.FileMode(mode & 0777)
-	switch mode & S_IFMT {
-	case syscall.S_IFBLK:
-		fm |= os.ModeDevice
-	case syscall.S_IFCHR:
-		fm |= os.ModeDevice | os.ModeCharDevice
-	case syscall.S_IFDIR:
-		fm |= os.ModeDir
-	case syscall.S_IFIFO:
-		fm |= os.ModeNamedPipe
-	case syscall.S_IFLNK:
-		fm |= os.ModeSymlink
-	case syscall.S_IFREG:
+func toFileMode(mode sshfx.FileMode) os.FileMode {
+	var ret = os.FileMode(mode & sshfx.ModePerm)
+
+	switch mode & sshfx.ModeType {
+	case sshfx.ModeNamedPipe:
+		ret |= os.ModeNamedPipe
+	case sshfx.ModeCharDevice:
+		ret |= os.ModeDevice | os.ModeCharDevice
+	case sshfx.ModeDir:
+		ret |= os.ModeDir
+	case sshfx.ModeDevice:
+		ret |= os.ModeDevice
+	case sshfx.ModeRegular:
 		// nothing to do
-	case syscall.S_IFSOCK:
-		fm |= os.ModeSocket
+	case sshfx.ModeSymlink:
+		ret |= os.ModeSymlink
+	case sshfx.ModeSocket:
+		ret |= os.ModeSocket
+	default:
+		ret |= os.ModeIrregular
```

Comment:
Hm, this might not be around in the minimum version we’re supporting?

As of go1.16, the good news is that they do actually define the `fs.ModeSetuid` `fs.ModeSetguid` and `fs.ModeSticky` flags for all architectures, regardless of their support. That should fortunately simplify a lot of this code.

Reply:
This was introduced in 1.11: https://golang.org/doc/go1.11#os